SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/PredictedGameplayCuesTest/CuesGASTestActor.cpp

```diff
@@ -0,0 +1,39 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "CuesGASTestActor.h"
+#include "GA_ApplyCueEffect.h"
+
+ACuesGASTestActor::ACuesGASTestActor()
+	: OnActiveCounter(0)
+	, ExecuteCounter(0)
+{
+}
+
+void ACuesGASTestActor::OnRep_Owner()
+{
+	// Make sure that any cached state in the ASC is up to date about our ownership chain.
+	// This is relevant for ability prediction - one of the conditions for allowing ability prediction is
+	// that the ASC's avatar actor is owned by the client's player controller.
+	// (See FGameplayAbilityActorInfo::IsLocallyControlled, which is checked before allowing ability prediction)
+	GetAbilitySystemComponent()->RefreshAbilityActorInfo();
+}
+
+void ACuesGASTestActor::OnActorReady(bool bHasAuthority)
+{
+	if (bHasAuthority)
+	{
+		// We need this actor to be an autonomous proxy on the owning client, since that is one of the conditions
```

Comment:
That's an odd hack to do, is it not possible to run the test on a PC-controlled actor?
The situation you are creating is unlikely to be reproduced in production, right?

Reply:
Yea I struggled with this - to avoid the hack, we'd want the actor with the ASC to be a pawn, so we can possess it witht he controller, which handles all the autonomous proxy stuff for us.
The reason I didn't do this yet was mostly convenience and code duplication. Namely, to make the target actor a pawn I'd have to:
1. Copy/paste all the ASC setup logic in the `AReplicatedGASTestActor`, since the new class has to inherit from `APawn`.
2. Do some additional possession tracking logic in the test implementation, since going by the existing possession test, you have to re-possess the original pawns at the end of the test. Though admittedly I'm not sure if this is a hard requirement.
https://github.com/spatialos/UnrealGDK/blob/509b683298967f3da1b8fc8303b8e717e7014b8b/SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/SpatialTestPossession.cpp#L82-L88

Reply:
You could also make a different game mode for this test and get a valid pawn by default.

Reply:
Good point, though according to Miron if a test is capable of running without dependence on configuration in the map, it should (so that we can run as many tests as possible in the same map). This also explains why the possession test cleans up after itself in that way.

Reply:
Okay, converting this to a pawn wasn't actually too bad. We still have some code duplication but it seems worth it for making it an actually realistic test case.